File: linked_list.py

class Node:
	def __init__(self, value):
		self.value = value
		self.next = None

class LinkedList:
	def __init__(self):
		self.head = None
	def append(self, value):
		if self.head is None:
			self.head = Node(value)
			return
		node = self.head
		while node.next:
			node = node.next
		node.next = Node(value)
		return
	def prepend(self, value):
		if self.head is None:
			self.head = Node(value)
			return
		node = Node(value)
		node.next = self.head
		self.head = node
		return
	def print(self):
		node = self.head
		while node:
			print(node.value)
			node = node.next
		return

	def pop(self):
		if self.head is None:
			return None
		node = self.head
		self.head = self.head.next
		return node.value
	
	def popleft(self):
		if self.head is None:
			return None
		node = self.head
		self.head = self.head.next
		return node.value
	
	def __len__(self):
		len = 0
		node = self.head
		while node:
			len += 1
			node = node.next
		return len
	
	def __getitem__(self, index):
		if index >= len(self):
			raise IndexError("Index out of range")
		node = self.head
		for i in range(index):
			node = node.next
		return node.value

	def __str__(self):
		node = self.head
		string = ""
		while node:
			string += str(node.value) + "->"
			node = node.next
		return string


linked_list = LinkedList()
linked_list.append(1)
linked_list.append(2)
linked_list.append(3)
linked_list.prepend(4)
print(linked_list)

while True:
	print("1. Append")
	print("2. Prepend")
	print("3. Pop")
	print("4. Pop Left")
	print("5. Length")
	print("6. Get Item")
	print("7. Print")
	print("8. Exit")
	choice = int(input("Enter your choice: "))

	
	if choice == 1:
		value = int(input("Enter value: "))
		linked_list.append(value)

	elif choice == 2:
		value = int(input("Enter value: "))
		linked_list.prepend(value)

	elif choice == 3:
		linked_list.pop()
	elif choice == 4:
		linked_list.popleft()
	elif choice == 5:
		print(len(linked_list))
	elif choice == 6:
		index = int(input("Enter index: "))
		print(linked_list[index])
	elif choice == 7:
		linked_list.print()
	elif choice == 8:
		break

File: linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Link
{
	int data;
	struct Link *next;
};

typedef struct Link *Link;

Link head = NULL;

void append(int data)
{
	Link newLink = (Link)malloc(sizeof(struct Link));
	newLink->data = data;
	newLink->next = NULL;

	if (head == NULL)
	{
		head = newLink;
		return;
	}

	Link current = head;
	while (current->next != NULL)
	{
		current = current->next;
	}
	current->next = newLink;
}

void prepend(int data)
{
	Link newLink = (Link)malloc(sizeof(struct Link));
	newLink->data = data;
	newLink->next = head;
	head = newLink;
}

void pop()
{
	if (head == NULL)
	{
		printf("List is empty\n");
		return;
	}

	Link current = head;
	while (current->next->next != NULL)
	{
		current = current->next;
	}
	free(current->next);
	current->next = NULL;
}

void popleft()
{
	if (head == NULL)
	{
		printf("List is empty\n");
		return;
	}

	Link temp = head;
	head = head->next;
	free(temp);
}

void nodeCount()
{
	int count = 0;
	Link current = head;
	while (current != NULL)
	{
		count++;
		current = current->next;
	}
	printf("Node count: %d\n", count);
}

void addPositional(int data, int position)
{
	Link newLink = (Link)malloc(sizeof(struct Link));
	newLink->data = data;
	newLink->next = NULL;

	if (head == NULL)
	{
		head = newLink;
		return;
	}

	Link current = head;
	int count = 0;
	while (current->next != NULL)
	{
		if (count == position-1)
		{
			newLink->next = current->next;
			current->next = newLink;
			return;
		}
		count++;
		current = current->next;
	}
}

void print()
{
	Link current = head;
	while (current != NULL)
	{
		printf("%d->", current->data);
		current = current->next;
	}
	printf("\n");
}

void dispMenu() {
	int data, choice;
	printf("1. Append\n");
	printf("2. Prepend\n");
	printf("3. Print\n");
	printf("4. Pop\n");
	printf("5. Pop Left\n");
	printf("6. Node Count\n");
	printf("7. Add Positional\n");
	printf("9. Exit\n");
	printf("Enter choice: ");
	scanf("%d", &choice);
	printf("\n");
	switch (choice)
	{
	case 1:
		printf("Enter data: ");
		scanf("%d", &data);
		append(data);
		break;

	case 2:
		printf("Enter data: ");
		scanf("%d", &data);
		prepend(data);
		break;
	case 3:
		print();
		break;
	case 4:
		pop();
		break;
	case 5:
		popleft();
		break;
	case 6:
		nodeCount();
		break;
	case 7:
		printf("Enter data: ");
		scanf("%d", &data);
		printf("Enter position: ");
		scanf("%d", &choice);
		addPositional(data, choice);
		break;
	case 9:
		exit(0);
	
	default:
		break;
	}
}

int main() {
	while (1) {
		dispMenu();
	}
	return 0;
}

